### PR TITLE
powerpc support

### PR DESCRIPTION
**Pull request by richo**
Originally, this was going to be discussed and revisted, however I've been working on this for months, and a rebase on top of master was about 1 flight's worth of work so I just went ahead and did it.

This gets you as far as being able to target powerpc with, eg:

```
LD_LIBRARY_PATH=./x86_64-unknown-linux-gnu/stage2/lib/ x86_64-unknown-linux-gnu/stage2/bin/rustc -C linker=powerpc-linux-gnu-gcc --target powerpc-unknown-linux-gnu hello.rs
```

Would really love to get this out before 1.0. r? @alexcrichton

### CODE REVIEW

File: src/libcoretest/mem.rs [hunk 1]
```diff
@@ -55,7 +56,8 @@ fn align_of_basic() {
 #[cfg(any(target_arch = "x86",
           target_arch = "arm",
           target_arch = "mips",
-          target_arch = "mipsel"))]
+          target_arch = "mipsel",
+          target_arch = "powerpc"))]
```
Comment (alexcrichton) [inline src/libcoretest/mem.rs]: Hm, to stem the flow of these additional `target_arch` settings, can you change this to be conditional on `target_pointer_width`?

File: src/librustc_trans/trans/asm.rs [hunk 2]
```diff
@@ -164,7 +164,8 @@ pub fn trans_inline_asm<'blk, 'tcx>(bcx: Block<'blk, 'tcx>, ia: &ast::InlineAsm)
 #[cfg(any(target_arch = "arm",
           target_arch = "aarch64",
           target_arch = "mips",
-          target_arch = "mipsel"))]
+          target_arch = "mipsel",
+          target_arch = "powerpc"))]
```
Comment (alexcrichton) [inline src/librustc_trans/trans/asm.rs]: Could you rewrite this to `not(condition)` where `condition` is what's below as well?

File: src/libcoretest/mem.rs (hunk 1 above)
Comment (richo) [inline src/libcoretest/mem.rs]: Just so we're on the same page, do this for all arches?

Comment (richo) [inline src/libcoretest/mem.rs]: Sorry. Ignore that other comment if it got emailed. Willfix.

File: src/libstd/rand/os.rs [hunk 3]
```diff
@@ -32,7 +32,8 @@ mod imp {
               any(target_arch = "x86_64",
                   target_arch = "x86",
                   target_arch = "arm",
-                  target_arch = "aarch64")))]
+                  target_arch = "aarch64",
+                  target_arch = "powerpc")))]
```
Comment (alexcrichton) [inline src/libstd/rand/os.rs]: Hm, this looks like... all architectures? Perhaps we could just list use linux here?

**Comment by alexcrichton** [comment]
This looks pretty good to me, thanks!

I think we're definitely growing a real need for an extension to `#[cfg]` which allows us to cut down on the huge amount of `#[cfg]`, but we can always worry about that later.

File: mk/main.mk [hunk 4]
```diff
@@ -263,7 +263,7 @@ endif
 ######################################################################
 
 # FIXME: x86-ism
-LLVM_COMPONENTS=x86 arm aarch64 mips ipo bitreader bitwriter linker asmparser mcjit \
+LLVM_COMPONENTS=x86 arm aarch64 mips powerpc ipo bitreader bitwriter linker asmparser mcjit \
```
Comment (alexcrichton) [inline mk/main.mk]: Due to this I think you'll also want to touch [the llvm-auto-clean-trigger](https://github.com/rust-lang/rust/blob/master/src/rustllvm/llvm-auto-clean-trigger) to ensure LLVM is rebuilt.

**Comment by alexcrichton** [comment]
As I also requested on https://github.com/rust-lang/rust/pull/19790, could you analyze the before/after sizes of the compiler? Just want to make sure LLVM isn't bringing too many megabyte wads of powerpc support!

**Comment by richo** [comment]
Awesome! Thanks so much for taking a look. I'll try to get these issues fixed tonight.

Regarding the size of the compiler:

After:
35M     x86_64-unknown-linux-gnu/stage2/lib/librustc_llvm-4e7c5e5c.so
12K     x86_64-unknown-linux-gnu/stage2/bin/rustc
3.2M   x86_64-unknown-linux-gnu/llvm/Release+Asserts/lib/libLLVMPower*.a

I don't have a before handy, but I can build one before I get home. Comparison numbers from that PR:

Before:
x86_64-unknown-linux-gnu/llvm/Release+Asserts/lib/libLLVMAArch64*.a: 0M
x86_64-unknown-linux-gnu/stage2/lib/librustc_llvm-4e7c5e5c.so: 36,259K
x86_64-unknown-linux-gnu/stage3/bin/rustc: 51,329K

(I'm not sure what the huge discrepancy with the stage3 rustc and my stage2 are, did a bunch of things move out into libs?)

File: src/libstd/rand/os.rs (hunk 3 above)
Comment (richo) [inline src/libstd/rand/os.rs]: This isn't mips or mipsel, which afaik doesn't have a getrandom(2) yet.

**Comment by richo** [comment]
I cleaned up the issues in your comments, and grabbed a nightly to check on the sizes of the objects:

From the nightly on rust-lang.org:

42M     ./lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_llvm-4e7c5e5c.so
8.0K    bin/rustc
(And obviously no powerpc archives).

I guess since I branched, some other LLVM changes must have landed? My librustc_llvm is 7 megs smaller.

**Comment by richo** [comment]
I finally made time to build the tree that I branched from, for comparison:

```
goodnight % for i in rust rust-vanilla; do du -hs $i/x86_64-unknown-linux-gnu/stage2/lib/librustc_llvm-*.so; done
42M     rust/x86_64-unknown-linux-gnu/stage2/lib/librustc_llvm-4e7c5e5c.so
40M     rust-vanilla/x86_64-unknown-linux-gnu/stage2/lib/librustc_llvm-4e7c5e5c.so
goodnight % for i in rust rust-vanilla; do du -hs $i/x86_64-unknown-linux-gnu/stage2/bin/rustc; done
8.0K    rust/x86_64-unknown-linux-gnu/stage2/bin/rustc
8.0K    rust-vanilla/x86_64-unknown-linux-gnu/stage2/bin/rustc
```

Is there anything else this needs for an r? @alexcrichton

**Comment by brson** [comment]
Nice.

**Pull request merged**